queue/src/main/java/org/killbill/queue/retry/RetryableSubscriber.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2014-2017 Groupon, Inc
+ * Copyright 2014-2017 The Billing Project, LLC
+ *
+ * The Billing Project licenses this file to you under the Apache License, version 2.0
+ * (the "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at:
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.killbill.queue.retry;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+import org.joda.time.DateTime;
+import org.killbill.bus.api.BusEvent;
+import org.killbill.clock.Clock;
+import org.killbill.notificationq.api.NotificationEvent;
+import org.killbill.notificationq.api.NotificationQueueService.NotificationQueueHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.reflect.TypeToken;
+
+public class RetryableSubscriber extends RetryableHandler {
+
+    private static final Logger log = LoggerFactory.getLogger(RetryableSubscriber.class);
+
+    public RetryableSubscriber(final Clock clock,
+                               final RetryableService retryableService,
+                               final NotificationQueueHandler handlerDelegate) {
+        super(clock, retryableService, handlerDelegate);
+    }
+
+    public void handleEvent(final BusEvent event) {
+        handleReadyNotification(new SubscriberNotificationEvent(event, event.getClass()),
+                                clock.getUTCNow(),
+                                event.getUserToken(),
+                                event.getSearchKey1(),
+                                event.getSearchKey2());
+    }
+
+    public interface SubscriberAction<T extends BusEvent> {
+
+        void run(T event);
+    }
+
+    public static final class SubscriberQueueHandler implements NotificationQueueHandler {
+
+        // Similar to com.google.common.eventbus.SubscriberRegistry
+        private static final LoadingCache<Class<?>, ImmutableSet<Class<?>>> FLATTEN_HIERARCHY_CACHE =
```

Comment:
If we were to only implement retry around `ExtBusEvent` (plugin only), then we would have only one type `org.killbill.billing.beatrix.extbus.DefaultBusExternalEvent`, and probably we could simplify quite a bit here

Reply:
Yes, probably, but we still want the retry piece for invoice plugins, no?